### PR TITLE
Capture stack traces while issuing IndexShard operations permits to easy debugging

### PR DESCRIPTION
**Pull request by bleskes**
Today we acquire a permit from the shard to coordinate between indexing operations, recoveries and other state transitions. When we leak an  permit it's practically impossible to find who the culprit is. This PR add stack traces capturing for each permit so we can identify which part of the code is responsible for acquiring the unreleased permit. This code is only active when assertions are active. 

The output is something like:
```
java.lang.AssertionError: shard [test][1] on node [node_s0] has pending operations:
--> java.lang.RuntimeException: something helpful 2
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:223)
	at org.elasticsearch.index.shard.IndexShard.<init>(IndexShard.java:322)
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:382)
	at org.elasticsearch.indices.IndicesService.createShard(IndicesService.java:514)
	at org.elasticsearch.indices.IndicesService.createShard(IndicesService.java:143)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.createShard(IndicesClusterStateService.java:552)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.createOrUpdateShards(IndicesClusterStateService.java:529)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:231)
	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateAppliers$6(ClusterApplierService.java:498)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:495)
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:482)
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:432)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:161)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:566)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:244)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:207)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)

--> java.lang.RuntimeException: something helpful
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:223)
	at org.elasticsearch.index.shard.IndexShard.<init>(IndexShard.java:311)
	at org.elasticsearch.index.IndexService.createShard(IndexService.java:382)
	at org.elasticsearch.indices.IndicesService.createShard(IndicesService.java:514)
	at org.elasticsearch.indices.IndicesService.createShard(IndicesService.java:143)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.createShard(IndicesClusterStateService.java:552)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.createOrUpdateShards(IndicesClusterStateService.java:529)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:231)
	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateAppliers$6(ClusterApplierService.java:498)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:495)
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:482)
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:432)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:161)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:566)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:244)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:207)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)

```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -2301,13 +2309,21 @@ public void onFailure(final Exception e) {
...
-                true);
+                true, debugInfo);
     }
 
     public int getActiveOperationsCount() {
         return indexShardOperationPermits.getActiveOperationsCount(); // refCount is incremented on successful acquire and decremented on close
     }
 
+    /**
+     * @return a list of containing an exceptions for each operation permit that wasn't released yet. The stack traces of the exceptions
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: there seems to be a missing word

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java [hunk 2]
```diff
@@ -53,10 +54,14 @@
...
     static final int TOTAL_PERMITS = Integer.MAX_VALUE;
     final Semaphore semaphore = new Semaphore(TOTAL_PERMITS, true); // fair to ensure a blocking thread is not starved
-    private final List<ActionListener<Releasable>> delayedOperations = new ArrayList<>(); // operations that are delayed
+    private final List<DelayedOperation> delayedOperations = new ArrayList<>(); // operations that are delayed
     private volatile boolean closed;
     private boolean delayed; // does not need to be volatile as all accesses are done under a lock on this
 
+    // only valid when assertions are enabled. Key is AtomicBoolean associated with each permit to ensure close once semantics. Value is an
+    // exception with some extra info in the message + a stack trace of the acquirer
+    private final Map<AtomicBoolean, RuntimeException> issuedPermits;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: any particular reason to use RuntimeException, and not just Throwable like in MockPageCacheRecycler

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java [hunk 3]
```diff
@@ -269,6 +305,28 @@ int getActiveOperationsCount() {
         }
     }
 
+    /**
+     * @return a list of containing an exceptions for each permit that wasn't released yet. The stack traces of the exceptions
+     *         was captured when the operation acquired the permit and their message contain the debug information supplied at the time.
+     */
+    List<RuntimeException> getActiveOperations() {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: can you test this method in IndexShardOperationPermitsTests that it returns something meaningful?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 4]
```diff
@@ -142,7 +142,7 @@ public RecoveryResponse recoverToTarget() throws IOException {
                 throw new DelayRecoveryException("source node does not have the shard listed in its state as allocated on the node");
             }
             assert targetShardRouting.initializing() : "expected recovery target to be initializing but was " + targetShardRouting;
-        });
+        }, shardId + " validating recovery target registered");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: add targetShardRouting??

**Review by ywelsch — APPROVED** [review]
LGTM. Left some nits

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: There's an unneeded s in exceptions and a missing s in contain. Can you clarify?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java (hunk 2 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: Jason scared me into never using Throwable. will switch.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java (hunk 3 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: yeah, so I was doubting about that one. What is meaningful? I can check there is something for each open op and also the message contains the debug info. Is that what you mean?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 4 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: sure

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: ok, I thought it meant to say "a list of **items/things/XYZ** containing an exception ..."

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java (hunk 3 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: yes. To check that the active operations are actually captured.

**Comment by bleskes** [comment]
@ywelsch I pushed some commits that addresses your feedback. Can you take another look?

**Review by ywelsch — APPROVED** [review]
LGTM. Thanks

File: server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java [hunk 5]
```diff
@@ -204,8 +214,24 @@ private void releaseDelayedOperations() {
...
+     *                        the tracing will capture the supplied object's {@link Object#toString()} value. Otherwise the object
+     *                        isn't used
+     *
      */
-    public void acquire(final ActionListener<Releasable> onAcquired, final String executorOnDelay, final boolean forceExecution) {
+    public void acquire(final ActionListener<Releasable> onAcquired, final String executorOnDelay, final boolean forceExecution,
+                        final Object debugInfo) {
+        final Throwable debugInfoWithStackTrace;
+        if (Assertions.ENABLED) {
+            debugInfoWithStackTrace = new Throwable(debugInfo.toString());
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java]: Is it possible to do this without a `Throwable`? Can you use a `Tuple` of `Object, StackTraceElement[]` populated with `Thread.currentThread().getStackTrace()` (and as an added bonanza avoid rendering the `Object` to a `String` unless absolutely needed)?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a suggestion.

**Comment by bleskes** [comment]
@jasontedor and I agreed to get this in now and follow up on removing the throwable.

**Pull request merged**